### DISCUSSION TITLE
is the --metershell blocking?

### DISCUSSION THREAD

**Discussion by naseef** [wmbusmeters/wmbusmeters#1421] [Q&A] [answered]
I want to use the `--metershell` option to run a shellscript. However the shellscript will take around 10 to 20 seconds to execute. Will this block the reception of  telegrams received during the scrips execution? or is it running on another thread? Please clarify

**Comment by weetmuts** [comment, marked as answer]
Yes, it definitely blocks the receiver thread. However, the serial port should have  buffer (https://forums.raspberrypi.com/viewtopic.php?t=222088) that should be enough for an occasional 10-20 seconds. However if you receive telegrams more often than that, then you have to have the shell push to some other queue systemet, otherwise the serial buffer will fill up.

There are only three threads. Read main.cc 
1) Main thread just waiting for everything to shut down.
2) Serial select /dev/ttyUSB0,1,2,3 etc, detect activitity trigger handler that reads from serial port, then invokes decoding, then invokes shells or other output.
3) Detector thread: detects changes in connected wmbus modules and check alarm conditions.

**Reply by naseef** [reply]
Thank you. 

If it is not too much to ask, is it possible to modify the shell running code to be executed in another thread?

**Reply by weetmuts** [reply]
Ah, no, way too much work for very little gain.  You can do it yourself with a bit if Unix-skill.
shell=/bin/sh -c "./dostuff.sh > the.logg 2>&1 &"

Put your code that takes time in dostuff.sh
